Improve concatenation analyzer

lucas-becirspahic opened the pull request:
Fix previous concatenation analyzer implementation :
* Don't trigger for concatenation used in property/field creation
* Trigger when an allocation is found (previously it would return an error only when the concatenation make more than 3 allocations (and ask to use a string builder instead))

farnyser commented on HotPathAllocationAnalyzer.Test/Analyzers/ConcatenationAllocationAnalyzerTests.cs:
```diff
@@ -8,18 +8,41 @@ namespace HotPathAllocationAnalyzer.Test.Analyzers
+            var sample = @"
+    public class Foo
+    {
+        public string Connection { get; set; } = ""User#"" + 0.ToString() + "" logged""; // should not allocate
```
I don't believe this change is necessary
As discussed previously, the issue we have is that the NoAllocation attribute scope is not working properly on properties.
The unit test does not show that, but in real usage, the analyzer should not even be called here.

farnyser commented:
Please extract the part about needing 3 concatenations => 1 so it can be merged, and see my comment about the other part not being needed. 🙏

Commit:
Fix concatenation analyzer

Trigger when an allocation is found (previously it would return an error only when the
concatenation make more than 3 allocations (and ask to use a string builder instead))
